Swift4 migration

DominicMDev opened the pull request:
Migrated to Swift 4.
Created a wrapper for RealmSwift.Results.

Commit:
Removed build errors in Xcode9

Commit:
Updated swift version file

Commit:
Updated podspec to version '1.1.0'

Commit:
Updated RealmSwift dependency to '3.0.0-beta'

Commit:
Updated dependency to 'RealmSwift' 3.0.0-beta

DominicMDev committed:
WIP: Creating `ResultsWrapper`

DominicMDev committed:
Created `ResultsWrapper`

DominicMDev committed:
Updated pods. Set swift version file to 4.0. Minor documentation fixes.

AnthonyMDev commented on QueryGenie/AttributeProtocol.swift:
```diff
@@ -535,7 +535,7 @@ public func << <A: AttributeProtocol, V>(left: A, right: [V]) -> NSComparisonPre
 }
 
-public func << <A: AttributeProtocol, V: Comparable>(left: A, right: Range<V>) -> NSComparisonPredicate where A.ValueType == V {
+public func << <A: AttributeProtocol, V>(left: A, right: Range<V>) -> NSComparisonPredicate where A.ValueType == V {
```
Why does `V` not need to conform to `Comparable` anymore? Though we are converting to `NSObject` in the `isInBetween` function, if they objects don't conform to `Comparable` the behavior for their comparison will be undefined, won't it?

JettF replied:
I think you're right. This was changed in commit dfed2ae. Am I wrong in thinking that when you made the xcode-9 branch you just spiked it? I, mistakenly, assumed that those changes were good to go. I will revisit those assumptions.

AnthonyMDev commented on QueryGenie/AttributeProtocol.swift:
```diff
@@ -35,7 +35,7 @@ public protocol NullableAttributeProtocol: AttributeProtocol {
 extension AttributeProtocol {
     
-    public final var ___expression: NSExpression {
+    public var ___expression: NSExpression {
```
For all of the below lines, is there a good reason for removing `final`? These functions should really not be overridden.

JettF replied:
I am not sure. I will try to figure it out. You removed them in commit dfed2ae.

JettF replied:
`final` was removed because the complier throws an error: "only classes and class members may be marked with 'final'. "

AnthonyMDev commented on QueryGenie/GenericQueryable.swift:
```diff
@@ -1,160 +0,0 @@
-import Foundation
-
-/// A query of items of a generic type `Element`.
-public protocol GenericQueryable: Queryable {
```
I like this, since `Queryable` is a `Sequence` we can just use the `Element` type of the sequence. It's already inherently "generic".

JettF replied:
Doh, that totally makes sense. Good catch.

AnthonyMDev commented on QueryGenie/CoreData/CoreDataQueryable.swift:
```diff
@@ -8,9 +8,7 @@ import Foundation
-public protocol CoreDataQueryable: GenericQueryable, Enumerable {
-    
-    associatedtype Element: NSFetchRequestResult
+public protocol CoreDataQueryable: Queryable, Enumerable where Element: NSFetchRequestResult {
```
Since we are getting rid of `GenericQueryable` this actually needs a new `associatedType` for the `NSFetchRequestResult`. The `Element` we are querying may not be the same type as the result.

JettF commented:
@AnthonyMDev If you would like me to adjust the formatting of the `ResultsWrapper` class to match our normal formatting, I can do that. We left it with the formatting from Realm so it would be easier to cross reference properties and methods in `Results`.

AnthonyMDev commented on QueryGenie/Realm/ResultsWrapper.swift:
```diff
@@ -0,0 +1,323 @@
+ - Note: `Results<Element>` allows `AnyObject`, setting the `Element` to `Object`
+         allows QueryGenie extensions to work with `Results`.
+*/
+public final class ResultsWrapper<T> where T: Object {
```
While this works, it does make the API less elegant. Now you have to constantly wrap your `Results` in this wrapper. Is there a way we could only make `Results` conform to our protocol when `Results.T: Object`?

It's been a while since I've gotten my head into this code base, but I feel like there should be a more elegant solution than this.

JettF replied:
I agree that there could be a more elegant solution. I will spend some more time thinking about this tomorrow. I would love to get away from the wrapper.

JettF replied:
If you want to take a crack at this, here are the problems we ran up against while developing the wrapper: (1) `Results` cannot be sub-classed. (2) Swift does not yet support conditional protocol conformance: https://github.com/apple/swift/blob/master/docs/GenericsManifesto.md

DominicMDev committed:
Removed `ResultsWrapper` and modified the constraint on `RealmQueryable`.

DominicMDev committed:
Removed White Space

Removed white space from Results+ResultsProtocol file.

DominicMDev committed:
Updated project build settings to recognize RealmSwift Module.
Also corrected paths to many of the frameworks that were being referenced incorrectly in the workspace.

DominicMDev committed:
Added ResultsQueryable to try to remove `objects()` from Results when Element isn't conforming to `Object`.
